### ISSUE TITLE
Use Custom SqlCLR Aggregate Functions in EF7

### ISSUE THREAD

**Issue by CorsairRO** [dotnet/efcore#28893]
## Ask a question

I have seen that Shay has added support for Aggregates in EF7 ( latest video in .NET Data Standup )

In that video it was a question from an attendee that asked how string.Join translate to STRINGAGG in versions of SQL Server Prior to 2016. The answer to that was NOT clear at all. Shay said it is NOT using STUFF instead of STRINGAGG, but he didnt explicitly said what happens if you use string.Join on a SQL Server older than 2016.

My apps has to support SQL Server back to 2012, therefore i have created myself a SqlCLR Aggregate, called dbo.StringConcantenate that all my queries in Sql are using when needing such a feature.

1) Can i call my SqlCLR function in any way in EF 7?
2) If not, can you please point me to a direction of how i can implement a custom translation to SQL Server 
string.Join => MyCLRAggregate instead of the default STRINGAGG
I DONT CARE about translating to other providers, i have always been using SQL Server since 2005.

Thank you very much
have a nice day

### Include provider and version information

EF Core version:
Database provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer)
Target framework: (e.g. .NET 6.0)
Operating system:
IDE: (e.g. Visual Studio 2022 17.3)

**Comment by roji** [comment]
@CorsairRO I think I misunderstood the question about STUFF in the community standup.

So first, with the EF 7 SQL Server provider, using string.Join always translates to STRINGAGG; this will simply fail on older versions of SQL Server where the function doesn't exist. This isn't worse than the EF 6 behavior, which throws a "cannot translate" exception when seeing string.Join.

Now, about the [STUFF function](https://docs.microsoft.com/en-us/sql/t-sql/functions/stuff-transact-sql?view=sql-server-ver16). STUFF does not really correspond to string.Join; it replaces a portion of a string with another string. More importantly, it is not an aggregate function, and so you cannot use it in aggregate contexts (e.g. in the projection of a query with GROUP BY). To fully understand this, consider the following two queries:

```c#
// Aggregate usage of string.Join. Concatenates all FirstNames for each category (an arbitrary number!) to a single string.
// The aggregate function (STRING_AGG) is called once per department, not once per row.
var allFirstNames = ctx.Blogs
    .GroupBy(b => b.Category)
    .Select(g => new {
        Category = g.Key,
        FirstNames = string.Join(", ", g.Select(b => b.FirstName))
    })
    .ToList();

// Non-aggregate usage of string.Join. Concatenates the FirstName and LastName of each individual row. Called once for each row.
_ = ctx.Blogs
    .Select(b => string.Join(", ", b.FirstName, b.LastName))
    .ToList();
```

If you're simply looking to do the second, you don't need STUFF. You can already write the following LINQ query:

```c#
_ = ctx.Blogs
    .Select(b => b.FirstName + ", " + b.LastName)
    .ToList();
```

This will do the exact same thing, and works well when the number of values to be concatenated is known in advance (not the case in the aggregate scenario).

Finally, SQL Server does have a [CONCAT_WS](https://docs.microsoft.com/en-us/sql/t-sql/functions/concat-ws-transact-sql?view=sql-server-ver16) function, which is more or less a non-aggregate version of STRING_AGG. It's only available since SQL Server 2017, but we should translate string.Join to it when it appears in non-aggregate contexts.

Hope the above clarifies everything let me know if you have further questions.

**Comment by CorsairRO** [comment]
Hello @roji 

Thank you for answering so fast...

- OK, i fully understand that it will be no difference with EF6 on .NET4 in the way that same error will be thrown.
- Also i know that STUFF is not exactly STRING_AGG, also STUFF is VERY inefficient.
That's why i have created my OWN aggregate as as SqlCLR Aggregate ( i dont know if you are aware that such functionality exists for more that a decade in SQL Server )
This was ( and still is ) the ONLY option to make efficient concatenate queries AND to support legacy SQL Server < 2016

In sql i have tons of queries like:
SELECT  b.BlogId, dbo.StringConcatenate( ', ', p.Title ) AS PostsTitleList
FROM blogs AS b
JOIN Posts as p ON b.BlogId = p.BlogId

That was my initial question about more or less. How can EF deal with Custom Aggregates...
OR, if it doesnt at all, if you have some startpoints to guide me to create my own sort of interceptor for string.Join to translate in SQL Server ONLY ( i dont care about any other DB ) to my custom aggregate instead of default to STRING_AGG
Or to add my own extensions to EF.Functions.
Or something else ...

PS: I have all my programs using .NET 4 moved to 5 and then to 6 without any issues, but because of the handicap with removing edmx support in EF Core i am still running EF6 ( plain old one ) for all my projects.
Now that you have added scaffolded with T4 in EF7 i will try to migrate all my data layers ( this was a BIG downside of EF Core since the beginning )
Also new interceptors you have also presented some time ago ( Especially object MATERIALZED ) helps to remove the stop on

I was thinking also to migrate lots of the SQL Logic into LINQ over EFCore, this is why i am asking about possible translation there because then i can remove 300+ objects from my every DB

Thank you again for improving the perf continously ( and, yes, of course, reduce ROUNDTRIPS )

**Comment by roji** [comment]
@CorsairRO mapping user aggregate functions is tracked by #27934, this isn't something that we currently support; please go vote on that, the infrastructure work done in 7.0 will probably make that issue relateively easy to implement in the future if enough users need it. We certainly haven't seen any demand for it up to now.

I don't think you'd be able to hack the above via an interceptor... However, you should be able to write a plugin with an aggregate method translator for your StringConcatenate function; that should unblock you. Writing a plugin isn't too hard, but we don't have any documentation for doing that. I'd recommend looking at the SQL Server NodaTime plugin, which has some aggregate translations ([here's the translator](https://github.com/dotnet/efcore/blob/release/7.0/src/EFCore.SqlServer.NTS/Query/Internal/SqlServerNetTopologySuiteAggregateMethodTranslator.cs)). Once you implement your aggregator and your plugin around it, you should be able to use your aggregate function.

**Comment by roji** [comment]
Duplicate of #27934
